connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java

```diff
@@ -582,4 +626,161 @@ public void onPartitionsRevoked(Collection<TopicPartition> partitions) {
+                    new Value());
+
+            offsetSeqNum = metricGroup.metrics().sensor("offset-seq-number");
+            offsetSeqNum.add(metricGroup.metricName("offset-commit-seq-no",
+                    "The current sequence number for offset commits"),
+                    new Value());
+
+            offsetCompletion = metricGroup.metrics().sensor("offset-commit-completion");
+            offsetCompletion.add(metricGroup.metricName("offset-commit-completion-rate",
+                    "The average per-second number of offset commit completions that were completed successfully."),
+                    new Rate());
+            offsetCompletion.add(metricGroup.metricName("offset-commit-completion-total",
+                    "The total number of offset commit completions that were completed successfully."),
+                    new Total());
+
+            offsetCompletionSkip = metricGroup.metrics().sensor("offset-commit-completion-skip");
+            offsetCompletionSkip.add(metricGroup.metricName("offset-commit-completion-skip-rate",
+                    "The average per-second number of offset commit completions that were received too late and skipped/ignored."),
+                    new Rate());
+            offsetCompletionSkip.add(metricGroup.metricName("offset-commit-completion-skip-total",
+                    "The total number of offset commit completions that were received too late and skipped/ignored."),
+                    new Total());
+
+            putBatchTime = metricGroup.metrics().sensor("put-batch-time");
+            putBatchTime.add(metricGroup.metricName("put-batch-max-time-ms",
+                    "The maximum time taken by this task to put a batch of sinks records."),
+                    new Max());
+            putBatchTime.add(metricGroup.metricName("put-batch-avg-time-ms",
+                    "The average time taken by this task to put a batch of sinks records."),
+                    new Avg());
+        }
+
+        void computeSinkRecordLag() {
+            Map<TopicPartition, OffsetAndMetadata> consumed = this.consumedOffsets.get();
+            Map<TopicPartition, OffsetAndMetadata> committed = this.committedOffsets.get();
+            long maxLag = 0L;
+            for (Map.Entry<TopicPartition, OffsetAndMetadata> committedOffsetEntry : committed.entrySet()) {
+                final TopicPartition partition = committedOffsetEntry.getKey();
+                final OffsetAndMetadata consumedOffsetMeta = consumed.get(partition);
+                if (consumedOffsetMeta != null) {
+                    final OffsetAndMetadata committedOffsetMeta = committedOffsetEntry.getValue();
+                    long consumedOffset = consumedOffsetMeta.offset();
+                    long committedOffset = committedOffsetMeta.offset();
+                    maxLag = Math.max(maxLag, consumedOffset - committedOffset);
```

Comment:
Not sure this is what people will generally mean by lag -- while the committed offset matters, normally if the consumer is in the process requesting the lag I think it'd mean the FetchRequest lag, i.e. how far behind *processing* the records is the consumer in comparison to what the broker indicates is the most recent offset.

in other words, I think i'd update this at the end of each `put()` and change from `committedOffsets` to `processedOffsets`.

Reply:
Okay, that's easy to change. But just to be clear, there are two things that lag could represent:

1. how many records the sink task has yet to process relative to the most recent offset written to the topic partitions, or
1. how many records the sink task has yet to process relative to the most recent offset read by the consumer.

Once I move from committedOffsets to processedOffsets as you suggest, the lag will be the second definition. Is that the preferred definition of lag?

Reply:
Hmm, good question. I may have actually been wrong about which values should be involved. I think @gwenshap and I had a long discussion about this awhile ago too and there are many ways you could define lag.

I think the real problem here is that we may not be exposing enough information from the consumer to compute what I would really think of as lag -- FetchRequests include high watermark info so you know how many records are in the log but not yet returned to you, and the consumer creates metrics based on that. But we don't have access to that info. A connector that commits on every message would look like it has 0 lag, but it could be very far behind in the topic.

Reply:
It's probably not a good thing performance-wise to get the end offsets from the consumer each time we poll, right?

Reply:
no, that'd be expensive as it is a network request per call. there might actually be a way to get at the info if it could be dug out of the metrics from the consumer, but that's obviously very much not ideal.

we actually might just need some info from users about what lag they think is important here too. they can get at the consumer lag themselves via a metric, so perhaps something connect-specific is fine. i just worry that they will see lag and assume it means essentially the same thing as it does for consumers. @gwenshap thoughts?

Reply:
Yes, the consumer metrics do capture the lag. I did try to carefully write the description of the metric:

> The maximum lag in terms of number of records behind the consumer the offset commits are for any topic partitions.

Hopefully that is clear enough, but I agree that it'd be better to have user feedback. Perhaps the metric name `sink-record-lag-max` is not ideal?